### PR TITLE
Make all istio-csr internal services into controller-runtime runnables.

### PR DESCRIPTION
**Pull request by JoshVanL**
This PR is branched from #69, which should be merged first.

This PR makes all internal services into controller-runtime runnables. This makes the code structure a lot cleaner, and means we have a single manager managing the long running packages.

Moves the readiness checks into controller-runtime.

Changes some of the chart values to reflect the new options structure.

/assign @jakexks

### CODE REVIEW

**Comment by JoshVanL** [comment]
Failed due to old version of Go. Updated to v1.6

/test all

**Review by irbekrm — CHANGES_REQUESTED** [review]
Looks good to me- the refactor makes sense.

I've read through the code and tried to deploy with a few options including with multiple replicas- I saw that although only one replica runs the rootCA configmap controller all of them serve the gRPC API which was something I wasn't sure about.

I've added a couple of minor nits, but happy to lgtm.

File: pkg/tls/tls.go [hunk 1]
```diff
@@ -59,111 +70,119 @@ type Provider struct {
...
+	// Before returning with the provider, we set a valid, up-to-date TLS
 	// config is ready for serving.
-	p.mustFetchCertificate(ctx)
-
-	go func() {
-		for {
-			// Create a new timer every loop. Renew 2/3 into certificate duration
-			renewalTime := (2 * p.opts.ServingCertificateDuration) / 3
-			timer := time.NewTimer(renewalTime)
+	notAfter, err := p.fetchCertificate(ctx)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/tls/tls.go:105]: Yes, that's right. I think this is fine, all errors which come from this function failing are likely configuration issues.

File: pkg/tls/tls.go [hunk 2]
```diff
@@ -59,111 +70,119 @@ type Provider struct {
...
+		if !notAfter.IsZero() {
+			p.log.Info("waiting to renew certificate", "renewal-time", time.Now().Add(renewalTime))
 		}
-	}()
 
-	p.readyz.Set(true)
+		select {
+		case <-ctx.Done():
+			p.log.Info("closing renewal", "context", ctx.Err())
+			timer.Stop()
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/tls/tls.go:124]: Yes, that sounds like a good addition.

File: cmd/app/app.go [hunk 3]
```diff
@@ -48,51 +52,73 @@ func NewCommand(ctx context.Context) *cobra.Command {
...
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to initialise cert-manager manager: %w", err)
 			}
 
-			// Create a new TLS provider for the serving certificate and private key.
-			tlsProvider, err := tls.NewProvider(ctx, opts.Logr, cm, readyz.Register(), opts.TLS)
+			intscheme := runtime.NewScheme()
+			if err := scheme.AddToScheme(intscheme); err != nil {
+				return fmt.Errorf("failed to add kubernetes scheme: %s", err)
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply cmd/app/app.go:62]: Why?

[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/app/app.go:62]: so that it's a string for '%s' 🤷🏼‍♀️

**Comment by irbekrm** [comment]
Looks good to me, thanks for making the changes!

/lgtm

**Pull request merged**